### PR TITLE
Ttahub 747/initial db migration

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-0

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/hooks/goal.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+          templateName: instance.name,
+          lastUsed: instance.createdAt,
+          regionId: grant.regionId,
+          creationMethod: 'Automatic',
+        },
+        transaction: options.transaction,
+      });
+    } catch (err) {
+      auditLogger.error(err);
+      auditLogger.error(JSON.stringify(err));
```
Comment (jasalisbury) [inline src/models/hooks/goal.js:24]: Can probably drop one of these log lines

Comment (jasalisbury) [inline src/models/hooks/goal.js:24]: Can probably drop one of these log lines

File: src/models/hooks/goal.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  }
+};
+
+const beforeValidate = async (sequelize, instance, options) => {
+  await autoPopulateGoalTemplateId(sequelize, instance, options);
+  autoPopulateOnApprovedAR(sequelize, instance);
+  preventNamChangeWhenOnApprovedAR(sequelize, instance);
+  autoPopulateStatusChangeDates(sequelize, instance);
+};
+// eslint-disable-next-line no-unused-vars
```
Comment (jasalisbury) [inline src/models/hooks/goal.js:113]: Can we remove this?

Comment (jasalisbury) [inline src/models/hooks/goal.js:113]: Can we remove this?

File: src/models/hooks/goalTemplate.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// import { auditLogger } from '../../logger';
```
Comment (jasalisbury) [inline src/models/hooks/goalTemplate.js:1]:
```suggestion
```

File: src/models/hooks/goalTemplate.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+const { Op } = require('sequelize');
+
+const autoPopulateHash = (sequelize, instance) => {
+  const changed = instance.changed();
+  if (Array.isArray(changed)
+    && changed.includes('templateName')
+    && instance.templateName !== null
+    && instance.templateName !== undefined) {
+    instance.set('hash', sequelize.fn('md5', sequelize.fn('TRIM', instance.templateName)));
```
Comment (jasalisbury) [inline src/models/hooks/goalTemplate.js:12]: Should we also lowercase/uppercase the `templateName` so `GOAL` AND `goal` have the same hash?

Comment (jasalisbury) [inline src/models/hooks/goalTemplate.js:12]: Should we also lowercase/uppercase the `templateName` so `GOAL` AND `goal` have the same hash?

File: src/models/hooks/goalTemplate.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  }
+};
+
+const autoPopulateCreationMethod = (sequelize, instance) => {
+  const changed = instance.changed();
+  if (Array.isArray(changed)
+        && (!changed.includes('creationMethod')
+        || instance.creationMethod === null
+        || instance.creationMethod === undefined)) {
+    instance.set('creationMethod', CREATION_METHOD[0]); // 'Automatic'
```
Comment (jasalisbury) [inline src/models/hooks/goalTemplate.js:32]: Could clean this up by defining the constants like

```
export const AUTOMATIC_CREATION = 'automatic'
export const CURATED_CREATION = 'curated'

export const CREATION_METHOD = [AUTOMATIC_CREATION, CURATED_CREATION]
```
that way we could use `AUTOMATIC_CREATION` here instead of the array index (might be better names for the string constants)

Comment (jasalisbury) [inline src/models/hooks/goalTemplate.js:32]: Could clean this up by defining the constants like

```
export const AUTOMATIC_CREATION = 'automatic'
export const CURATED_CREATION = 'curated'

export const CREATION_METHOD = [AUTOMATIC_CREATION, CURATED_CREATION]
```
that way we could use `AUTOMATIC_CREATION` here instead of the array index (might be better names for the string constants)

File: src/widgets/goalStatusGraph.test.js [hunk 6]
```diff
@@ -60,17 +60,17 @@ describe('goalStatusGraph', () => {
...
+    //   status: '',
+    //   grantId,
+    //   recipientId,
+    // }));
+
+    // goals.push(await createGoal({
+    //   status: null,
+    //   grantId,
+    //   recipientId,
+    // }));
```
Comment (jasalisbury) [inline src/widgets/goalStatusGraph.test.js:73]:
```suggestion
```

Comment (jasalisbury) [inline src/widgets/goalStatusGraph.test.js:73]:
```suggestion
```

**Review by jasalisbury — COMMENTED** [review]
About half way through reviewing, will hopefully get everything excluding the migration reviewed tomorrow

File: src/routes/files/handlers.js [hunk 7]
```diff
@@ -40,8 +50,65 @@ const altFileTypes = [
...
+      && objectiveId in file.objectiveFiles.map((r) => r.objectiveId)) {
+      const of = file.objectiveFiles.find((r) => r.objectiveId === objectiveId);
+      deleteObjectiveFile(of.id);
+    } else if (objectiveTempleteId
+      && objectiveTempleteId in file.objectiveTemplateFiles.map((r) => r.objectiveTempleteId)) {
+      const otf = file.objectiveTemplateFiles
+        .find((r) => r.objectiveTempleteId === objectiveTempleteId);
+      deleteObjectiveTemplateFile(otf.id);
+    }
+    file = await getFileById(fileId);
```
Comment (jasalisbury) [inline src/routes/files/handlers.js:89]: If you want to reload the file from the DB you should be able to just call `await file.reload()` (granted I'm not sure if it properly reloads any associations)

**Review by jasalisbury — COMMENTED** [review]
Another batch of comments. There are some things that need addressed in multiple places that I stopped commenting:

* Commented code needs to be removed
* `findOrCreate` calls with the same `default` and `where` clause don't need `default` specified (unless I'm missing something!)

File: src/models/hooks/activityReport.js [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+          attributes: ['objectiveId'],
+          where: { activityReportId: instance.id },
+          include: [
+            {
+              model: sequelize.models.Objective,
+              as: 'objectives',
+              attributes: ['goalId'],
+            },
+          ],
+          transaction: options.transaction,
```
[the remark this answers is not in the document]
Comment (GarrettEHill) [reply src/models/hooks/activityReport.js:158]: I dont know

File: src/models/hooks/goalTemplate.js (hunk 4 above)
Comment (GarrettEHill) [reply src/models/hooks/goalTemplate.js:12]: We are going to refine the comparison more in the next phase, but lower could prevent some corrections from being saved.

Comment (GarrettEHill) [reply src/models/hooks/goalTemplate.js:12]: We are going to refine the comparison more in the next phase, but lower could prevent some corrections from being saved.

File: src/models/hooks/objective.js [hunk 9]
```diff
@@ -0,0 +1,195 @@
+// import { Op } from 'sequelize';
+import { auditLogger } from '../../logger';
+
+const autoPopulateObjectiveTemplateId = async (sequelize, instance, options) => {
+  // eslint-disable-next-line no-prototype-builtins
+  if (!instance.hasOwnProperty('objectiveTemplateId')
```
[the remark this answers is not in the document]
Comment (GarrettEHill) [reply src/models/hooks/objective.js:6]: So on the caller end do something link: autoPopulateObjectiveTemplateId({sequelize, instance, options});
Then on the implementation it would be autoPopulateObjectiveTemplateId(req), then use would be req.sequelize.

File: src/models/hooks/objective.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+          [regionId] = otherEntity.regionIds;
+        }
+      } catch (err) {
+        auditLogger.error(JSON.stringify(err));
+        throw err;
+      }
+    }
+
+    const objectiveTemplate = await sequelize.models.ObjectiveTemplate.findOrCreate({
+      where: { hash: sequelize.fn('md5', sequelize.fn('NULLIF', sequelize.fn('TRIM', instance.title), '')), regionId },
```
[the remark this answers is not in the document]
Comment (GarrettEHill) [reply src/models/hooks/objective.js:54]: So you want to gen the md5 in js instead of having the db use the standard md5 function built into the db? This is a distinct way to reproducibly generate md5s.

[the remark this answers is not in the document]
Comment (jasalisbury) [reply src/models/hooks/objective.js:54]: I'm saying we move the text that gets sent to the DB to generate md5s into it's own function. That way what we pass to the DB is defined in one place

File: src/routes/files/handlers.test.js [hunk 11]
```diff
@@ -3,6 +3,7 @@ import waitFor from 'wait-for-expect';
 import db, {
```
[the remark this answers is not in the document]
Comment (GarrettEHill) [reply src/routes/files/handlers.test.js:3]: yes

File: src/models/hooks/objective.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jasalisbury) [reply src/models/hooks/objective.js:6]: I'm suggesting adding a function like:

```
function isDefined(object, prop) {
  return !object.hasOwnProperty(prop) || instance[prop] === null || instance[prop] === undefined
}
```

Then this line could be
```
if isDefined(instance, 'objectiveTemplateId') {
```

[the remark this answers is not in the document]
Comment (jasalisbury) [reply src/models/hooks/objective.js:6]: because it seems there are some other places the new "isDefined" function could be used (might be a better name for it too!)

File: src/seeders/20210127161802-add-goals.js [hunk 12]
```diff
@@ -12,82 +12,99 @@ const bulletedGoal = `
...
+        grantId: 2,
+        onApprovedAR: false,
+      },
+    ];
+
+    await queryInterface.bulkInsert('GoalTemplates', goalTemplates, { validate: true, individualHooks: true });
+    await queryInterface.bulkInsert('Goals', goals, { validate: true, individualHooks: true });
+
+    await queryInterface.sequelize.query(`ALTER SEQUENCE "GoalTemplates_id_seq" RESTART WITH ${goalTemplates[goalTemplates.length - 1].id + 1};`);
+    await queryInterface.sequelize.query(`ALTER SEQUENCE "Goals_id_seq" RESTART WITH ${goals[goals.length - 1].id + 1};`);
```
Comment (jasalisbury) [inline src/seeders/20210127161802-add-goals.js:103]: Seems weird that we have to do this. I assume there are bugs without these lines. Do you know why we have to restart the goal and goal template id sequences?

File: src/services/activityReports.js [hunk 13]
```diff
@@ -184,7 +186,30 @@ async function saveReportRecipients(
...
+      if (newRecipient.otherEntityId) {
+        return ActivityRecipient.findOrCreate({
+          where: {
+            activityReportId: newRecipient.activityReportId,
+            otherEntityId: newRecipient.otherEntityId,
+          },
+          defaults: newRecipient,
+        });
+      }
+      return null;
```
Comment (jasalisbury) [inline src/services/activityReports.js:209]: We will want to remove nulls, with a filter before the map looking for recipients with no `grantId` or `otherEntityId` or after looking for nulls

File: src/services/activityReports.test.js [hunk 14]
```diff
@@ -492,7 +493,13 @@ describe('Activity report service', () => {
           recipientNextSteps: [{ note: 'One Piece' }, { note: 'Toy Story' }],
         };
         // When that report is created
-        const report = await createOrUpdate(reportObjectWithNotes);
+        let report;
+        try {
```
Comment (jasalisbury) [inline src/services/activityReports.test.js:497]: Can we change mult-try catch into a single try catch?

File: src/services/objectives.js [hunk 15]
```diff
@@ -6,6 +6,7 @@ import {
 import { removeUnusedGoalsObjectivesFromReport } from './goals';
 
 export async function saveObjectivesForReport(objectives, report) {
+  // TODO: ttaProvided needs to move from objective to ActivityReportObjective
```
Comment (jasalisbury) [inline src/services/objectives.js:9]: Should we make this change in this PR?

File: src/services/goalServices/destroyGoal.test.js [hunk 16]
```diff
@@ -40,17 +39,17 @@ describe.skip('destroyGoal handler', () => {
       status: 'Not Started',
     });
 
-    await GrantGoal.create({
-      recipientId: recipient.id,
-      grantId: grant.id,
-      goalId: goal.id,
-    });
+    // await GrantGoal.create({
```
Comment (jasalisbury) [inline src/services/goalServices/destroyGoal.test.js:42]: Commented code

File: src/services/goals.js [hunk 17]
```diff
@@ -8,12 +8,13 @@ import {
   ObjectiveResource,
   ObjectiveTopic,
   ActivityReportObjective,
-  GrantGoal,
+  // GrantGoal,
```
Comment (jasalisbury) [inline src/services/goals.js:11]: commented code

File: src/services/goals.js [hunk 18]
```diff
@@ -137,6 +139,9 @@ export async function goalByIdWithActivityReportsAndRegions(goalId) {
   });
 }
 
+// eslint-disable-next-line no-empty-function
+export async function copyGoalsToGrants(goals, grantIds) {}
```
Comment (jasalisbury) [inline src/services/goals.js:143]: If we aren't using this function can we remove it?

File: src/services/goals.js [hunk 19]
```diff
@@ -194,141 +199,137 @@ async function cleanupObjectivesForGoal(goalId, currentObjectives) {
...
+      id,
+      grantId,
+      recipientId,
+      regionId,
+      objectives,
+      ...fields
+    } = goalData;
+
+    // there can only be one on the goal form (multiple grants maybe, but one recipient)
+    recipient = recipientId; // TODO: this is wrong
```
Comment (jasalisbury) [inline src/services/goals.js:221]: Can we make this not wrong?

File: src/services/goals.js [hunk 20]
```diff
@@ -194,141 +199,137 @@ async function cleanupObjectivesForGoal(goalId, currentObjectives) {
...
-  // // there can only be one on the goal form (multiple grants maybe, but one recipient)
-  // // we will need this after the transaction, as trying to do a find all within a transaction
-  // // yields the previous data values
-  // let recipient;
+  // there can only be one on the goal form (multiple grants maybe, but one recipient)
+  // we will need this after the transaction, as trying to do a find all within a transaction
+  // yields the previous data values
+  let recipient;
 
   // eslint-disable-next-line max-len
```
Comment (jasalisbury) [inline src/services/goals.js:209]: Doubt we need this anymore

File: src/services/goals.js [hunk 21]
```diff
@@ -194,141 +199,137 @@ async function cleanupObjectivesForGoal(goalId, currentObjectives) {
...
-  //         where: {
-  //           id: {
-  //             [Op.notIn]: objectiveResources.length
-  //               ? objectiveResources.map(([or]) => or.id) : [],
-  //           },
-  //           objectiveId: objective.id,
-  //         },
-  //       });
+        // cleanup objective topics
+        await ObjectiveTopic.destroy({
```
Comment (jasalisbury) [inline src/services/goals.js:284]: Can we move any objective cleanup to the `cleanupObjectivesForGoals` function?

File: src/services/goals.js [hunk 22]
```diff
@@ -194,141 +199,137 @@ async function cleanupObjectivesForGoal(goalId, currentObjectives) {
...
-  //   return goal.id;
-  // })));
+    return newGoal.id;
+  }));
 
-  // // we have to do this outside of the transaction otherwise
-  // // we get the old values
-  // return goalsByIdAndRecipient(goalIds, recipient);
+  // we have to do this outside of the transaction otherwise
+  // we get the old values
```
Comment (jasalisbury) [inline src/services/goals.js:331]: This comment is no longer true because `goalsByIdAndRecipient` takes place in the same transaction as all the queries above

File: src/services/goals.js [hunk 23]
```diff
@@ -391,6 +392,7 @@ async function removeActivityReportObjectivesFromReport(reportId, objectiveIdsTo
   });
 }
 
+// TODO: ask Josh what the intent of this is for
```
Comment (jasalisbury) [inline src/services/goals.js:395]: `removeUnusedGoalsObjectivesFromReport` calls this function to remove any goals on activity reports that are no longer in use. A user adds a goal to an AR, saves, and then removes that goal from the AR and saves again.

File: src/services/goals.js [hunk 24]
```diff
@@ -466,27 +468,53 @@ export async function saveGoalsForReport(goals, report) {
...
     }
 
-    // using upsert here and below
-    // - add returning: true to options to get an array of [<Model>,<created>] (postgres only)
-    // - https://sequelize.org/v5/class/lib/model.js~Model.html#static-method-upsert
-    const [newGoal] = await Goal.upsert(fields, { returning: true });
+    // This linkage of goal directly to a report will allow a save to be made even if no objective
+    // has been started.
+    // A hook on ActivityReportGoal will automatically update the status of the goal if required.
+    await ActivityReportGoal.findOrCreate({
```
Comment (jasalisbury) [inline src/services/goals.js:492]: I don't remember seeing any place (might have just missed it!) where this linkage is undone. If I add a goal to a report and then decide I don't want that goal on my report

**Review by jasalisbury — COMMENTED** [review]
Have everything reviewed up to the migration `src/migrations/20220425000000-goal-templates.js`. If you work on the goal service I'd check if Matt already has updates, I just saw a commit hit his branch referencing the goals service.

**Comment by thewatermethod** [comment]
@GarrettEHill , @jasalisbury  
FYI - I had to substantially rewrite the ```saveGoalsForReport``` function in my branch to get it working.

**Review by jasalisbury — COMMENTED** [review]
Found another file I needed to review

File: src/migrations/20220425000000-goal-templates.js [hunk 25]
```diff
@@ -0,0 +1,1964 @@
...
+          `DO $$
+          BEGIN
+            CREATE TEMP TABLE "TempRoles" AS
+            SELECT
+              MIN(r."id") "id",
+              r.name,
+              r."fullName",
+              MIN(r."createdAt") "createdAt",
+              MAX(r."updatedAt") "updatedAt",
+              MAX(r."updatedAt") "deletedAt",
```
Comment (jasalisbury) [inline src/migrations/20220425000000-goal-templates.js:451]: Is this filling the `deletedAt` field in the new table as the max `updatedAt`? Wouldn't we want the new table to have a null `deletedAt` by default?

File: src/migrations/20220425000000-goal-templates.js [hunk 26]
```diff
@@ -0,0 +1,1964 @@
...
+        );
+      } catch (err) {
+        console.error(err); // eslint-disable-line no-console
+        throw (err);
+      }
+    },
+  ),
+  down: async (queryInterface) => queryInterface.sequelize.transaction(
+    async (transaction) => {
+      await queryInterface.dropTable('GoalTemplates', { transaction });
```
Comment (jasalisbury) [inline src/migrations/20220425000000-goal-templates.js:1960]: Should we throw an error if `down` is ran? With how big this migration is I doubt we will be able to directly roll back, so might as well error out and tell the person running the migration to reload the DB from a backup if they want to roll back.

**Comment by kryswisnaskas** [comment]
I don't see the duplicates bug anymore! 👍

File: src/models/hooks/activityReportCollaborator.js [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+  if (!roles.length || roles.length === 0) {
+    return user.fullName;
+  }
+  const combinedRoles = roles.reduce((result, val) => {
+    if (val) {
+      return val === 'TTAC' || val === 'COR' ? `${result}, ${val}` : `${result}, ${val.split(' ').map((word) => word[0]).join('')}`;
+    }
+    return '';
+  }, []);
+  return combinedRoles.length > 0 ? `${user.name}${combinedRoles}` : user.name;
```
Comment (jasalisbury) [inline src/models/hooks/activityReportCollaborator.js:12]: Do we need a space between `user.name` and `combinedRoles`?

File: src/scopes/goals/index.js [hunk 28]
```diff
@@ -6,14 +6,15 @@ import { withTopics, withoutTopics } from './topics';
...
+import { withRoles, withoutRoles } from './role';
 import { withGrantNumber, withoutGrantNumber } from './grantNumber';
 
 export const topicToQuery = {
   createDate: {
     bef: (query) => beforeCreateDate(query),
     aft: (query) => afterCreateDate(query),
     win: (query) => withinCreateDate(query),
-    is: (query) => withinCreateDate(query),
+    in: (query) => withinCreateDate(query),
```
Comment (jasalisbury) [inline src/scopes/goals/index.js:17]: Was this change intentional?

Comment (thewatermethod) [reply src/scopes/goals/index.js:17]: Yeah, all the other date scopes were "in", and the frontend was sending an "in"

File: src/models/hooks/activityReportCollaborator.js (hunk 27 above)
Comment (thewatermethod) [reply src/models/hooks/activityReportCollaborator.js:12]: combinedRoles starts with a comma, so this concatenates to things like ```Cecil Koss, ECS```

File: src/migrations/20220425000000-goal-templates.js [hunk 29]
```diff
@@ -0,0 +1,1964 @@
...
+          type: Sequelize.BOOLEAN,
+          allowNull: true,
+          default: false,
+        }, { transaction });
+
+        await queryInterface.sequelize.query(
+          `DO $$
+          BEGIN
+          ------------------------------------------------------------------------------------
+          UPDATE ONLY "Goals"
```
Comment (hardwarehuman) [inline src/migrations/20220425000000-goal-templates.js:621]: I think you meant to update `Objectives` here

File: src/models/hooks/goal.js [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+  }
+};
+
+const preventNamChangeWhenOnApprovedAR = (sequelize, instance) => {
+  if (instance.onApprovedAR === true) {
+    const changed = instance.changed();
+    if (instance.id !== null
+      && Array.isArray(changed)
+      && changed.includes('name')) {
+      throw new Error('Goal name change now allowed for goals on approved activity reports.');
```
Comment (kryswisnaskas) [inline src/models/hooks/goal.js:53]:
```suggestion
      throw new Error('Goal name change not allowed for goals on approved activity reports.');
```

**Review by hardwarehuman — COMMENTED** [review]
One more  incremental comment update

**Review by kryswisnaskas — APPROVED** [review]
Nice work! 👍
